asllib/doc/asl.spec

```
@@ -1295,10 +1297,3181 @@ typing function annotate_literal(tenv: static_envs, l: literal) -> (t: ty)
+         (res: list0(int_constraint)) | constants_set(CannotUnderapproximate,CannotOverapproximate)
+{
+  "filters the list of constraints {cs} for {op},
+  removing constraints that represents a division
+  operation that will definitely fail when {op} is the
+  division operation. If the list is empty, the result
+  is either $\CannotUnderapproximate$ or
+  $\CannotOverapproximate$, based on {approx} (this
+  function is invoked in the context of approximating
+  lists of constraints).",
+  prose_application = "",
+  math_layout = [_,_],
+};
+
+relation filter_reduce_constraint_div(c: int_constraint) ->
+         (c_opt: option(int_constraint))
+{
+  "returns $\None$ if {c} is an exact constraint for a
+  \binopexpressionterm{} for dividing two integer
+  literals where the denominator does not divide the
+  numerator, and an optional containing {c} otherwise.
+  The result is returned in {c_opt}. This is used to
+  conservatively test whether {c} would always fail
+  dynamically.",
+  prose_application = "",
+};
+
+relation get_literal_div_opt(e: expr) ->
+         (range_opt: option((Z, Z)))
+{
+  "matches the expression {e} to a binary operation
+  expression over the division operation and two literal
+  integer expressions. If {e} matches this pattern, the
+  result {range_opt} is an optional containing the pair
+  of integers appearing in the operand expressions.
+  Otherwise, the result is $\None$.",
+  prose_application = "",
+};
+
+relation explode_intervals(tenv: static_envs, cs: list0(int_constraint)) ->
+         (new_cs: list0(int_constraint), p: precision_loss_indicator)
+{
+  "applies $\explodeintervals$ to each constraint of {cs}
+    in {tenv}, and returns a pair consisting of the list
+    of exploded constraints in {new_cs} and a
+    \precisionlossindicatorterm{} {p}.",
+  prose_application = "",
+};
+
+relation interval_too_large(z1: Z, z2: Z) ->
+         (b: Bool)
+{
+  "determines whether the set of numbers between {z1} and
+  {z2}, inclusive, contains more than
+  $\maxexplodedintervalsize$ integers, yielding the
+  result in {b}.",
+  prose_application = "",
+};
+
+relation binop_is_exploding(op: binop) ->
+         (b: Bool)
+{
+  "determines whether the binary operation {op} should
+  lead to applying $\explodeintervals$ when the {op} is
+  applied to a pair of constraint lists. It is assumed
+  that {op} is one of $\MUL$, $\SHL$, $\POW$, $\ADD$,
+  $\DIV$, $\SUB$, $\MOD$, $\SHR$, and $\DIVRM$.",
+  prose_application = "",
+};
+
+relation mem_bfs(tenv: static_envs, bfs2: list1(bitfield), bf1: bitfield) ->
+         (b: Bool)
+{
+  "checks whether the bitfield {bf1} exists in {bfs2} in
+  the context of {tenv}, returning the result in {b}.",
+  prose_application = "",
+};
+
+relation to_well_constrained(t: ty) ->
+         (t': ty)
+{
+  "returns the \wellconstrainedversionterm{} of a type {t}
+  in {t'}, which converts
```

This is difficult to read, what about 


```suggestion
  "returns the \wellconstrainedversionterm{} {t'} of a type {t}, which converts
```

or
```suggestion
  "returns {t'}, the \wellconstrainedversionterm{} of a type {t}, which converts
```